apiserver/common/crossmodel/crossmodel.go

```diff
@@ -0,0 +1,185 @@
+		logger.Debugf("unit %v has departed relation %v", unitTag.Id(), relationTag.Id())
+		ru, err := rel.RemoteUnit(unitTag.Id())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		logger.Debugf("%s leaving scope", unitTag.Id())
+		if err := ru.LeaveScope(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	for _, change := range change.ChangedUnits {
+		unitTag := names.NewUnitTag(fmt.Sprintf("%s/%v", applicationTag.Id(), change.UnitId))
+		logger.Debugf("changed unit tag for remote id %v is %v", change.UnitId, unitTag)
+		ru, err := rel.RemoteUnit(unitTag.Id())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		inScope, err := ru.InScope()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		settings := make(map[string]interface{})
+		for k, v := range change.Settings {
+			settings[k] = v
+		}
+		if !inScope {
+			logger.Debugf("%s entering scope (%v)", unitTag.Id(), settings)
+			err = ru.EnterScope(settings)
+		} else {
+			logger.Debugf("%s updated settings (%v)", unitTag.Id(), settings)
+			err = ru.ReplaceSettings(settings)
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func getRemoteEntityTag(backend Backend, id params.RemoteEntityId) (names.Tag, error) {
+	modelTag := names.NewModelTag(id.ModelUUID)
+	return backend.GetRemoteEntity(modelTag, id.Token)
+}
+
+// WatchLocalRelationUnits returns a watcher for changes to the units on the
+// specified relation.
+func WatchLocalRelationUnits(backend Backend, tag names.RelationTag) (state.RelationUnitsWatcher, error) {
+	relation, err := backend.KeyRelation(tag.Id())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, ep := range relation.Endpoints() {
+		_, err := backend.Application(ep.ApplicationName)
+		if errors.IsNotFound(err) {
+			// Not found, so it's the remote application. Try the next endpoint.
+			continue
+		} else if err != nil {
+			return nil, errors.Trace(err)
+		}
+		w, err := relation.WatchUnits(ep.ApplicationName)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return w, nil
+	}
+	return nil, errors.NotFoundf("local application for %s", names.ReadableString(tag))
+}
+
+// RelationUnitSettings returns the unt settings for the specified relation unit.
+func RelationUnitSettings(backend Backend, ru params.RelationUnit) (params.Settings, error) {
+	relationTag, err := names.ParseRelationTag(ru.Relation)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	rel, err := backend.KeyRelation(relationTag.Id())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	unitTag, err := names.ParseUnitTag(ru.Unit)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	unit, err := rel.Unit(unitTag.Id())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	settings, err := unit.Settings()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	paramsSettings := make(params.Settings)
+	for k, v := range settings {
+		vString, ok := v.(string)
```

Comment:
Perhaps, but params.Settings has always been defined to be map[string]interface{}
I'm always wary of implicit typing via String() or whatever, especially for on the wire params. I think best to be explicit. We can revisit later, but it's a question for more than just cmr.